### PR TITLE
Added the C++11 changes back in. Replaces #327

### PR DESCRIPTION
**Pull request by OniDaito**
Ok so these are the C++11 changes as discussed I think. I've included all the changes you've made up this point. Is this better for you?

### CODE REVIEW

**Comment by OniDaito** [comment]
I've kept the 'doc' directory with the markdown in it, so hopefully we'll be able to integrate that later on.

**Comment by mloskot** [comment]
@OniDaito there was a few issues though, see my additional commits.

In case the build is failing (including SOCI_CXX_C11=OFF), feel free to submit additional PRs to correct things. I will try to add C++11 job to Travis CI, but AFAIS setting cutting-edge GCC there is PITA and clang seems to be a troublemaker with Oracle.

File: tests/db2/test-db2.cpp [hunk 1]
```diff
@@ -20,6 +20,17 @@ using namespace soci::tests;
 std::string connectString;
 backend_factory const &backEnd = *soci::factory_db2();
 
+// It appears later versions of GCC arent happy with this - to be fixed properly
+
+#if (__GNUC__ == 4 && (__GNUC_MINOR__ > 6)) || (__clang__ == 1)
+namespace boost {
+    std::basic_ostream<char, std::char_traits<char> >& operator<< (std::basic_ostream<char, std::char_traits<char> > & stream, boost::optional<int> const & value){
```
Comment (mloskot) [inline tests/db2/test-db2.cpp]: Why not inline operator in common-tests.h?

Comment (OniDaito) [inline tests/db2/test-db2.cpp]: Yeah, I was thinking that as I pasted them again. Should probably do that. It's neater. Cheers

Comment (mloskot) [inline tests/db2/test-db2.cpp]: No problem, it's just that kind of thing that raises questions :)
As I said, feel free to improve things with following PRs.
(I do really need to take care of Travis CI for C11 mode, so you have safety net too.)

Comment (OniDaito) [inline tests/db2/test-db2.cpp]: So is Clang being a real pain for Oracle then? Presumably we'll see this in the OS X tests as well?

Comment (mloskot) [inline tests/db2/test-db2.cpp]: I'm still unclear myself, so will have to look again before I know what's the problem :)

**Pull request merged**